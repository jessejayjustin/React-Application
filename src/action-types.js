export const GET_CHARITIES = 'GET_CHARITIES';
export const UPDATE_TOTAL_DONATIONS = 'UPDATE_TOTAL_DONATIONS';
export const UPDATE_MESSAGE = 'UPDATE_MESSAGE';
export const ERROR_MESSAGE = 'ERROR_MESSAGE';

