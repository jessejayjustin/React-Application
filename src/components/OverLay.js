import React, { Component } from "react";
import styled from 'styled-components';

class OverLay extends Component {
  constructor() {
    super();
    this.state = {
      selectedItem: [],
      selectedAmount: 10
    };
  }

  handlePay(id, amount, currency) {
    this.setState(
      {
        selectedItem: {
          id: id,
          amount: amount,
          currency: currency
        }
      },
      function() {
        this.props.handlePay(this.state.selectedItem);
      }
    );
  }

  selectedAmount(amount) {
    this.setState({ selectedAmount: amount })
  }

  render() {
  
    const {payments, item, hideOverlay} = this.props;

    const overLayStyle = {
      backgroundColor: 'white',
      transition: '0.5s',
      zIndex: '1',
      width: '530px',
      height: '320px',
      opacity: '0.9',
      position: 'absolute',
      left: '0px',
      top: '0px'
    }

    const inputStyle = {
      display: 'inline-block',
      float: 'left',
      width: '12%',
      position: 'relative',
      left: '115px',
      top: '135px',
    }

    const textStyle = {
      fontSize: '18px',
      color: 'rgb(96,96,96)',
      fontFamily: 'Trebuchet MS, Helvetica, sans-serif'
    }

    const payBtn = {
      position: 'absolute',
      top: '190px',
      left: '245px',
      padding: '5px 9px',
      backgroundColor: 'rgb(255,255,255)',
      color: 'rgb(30,144,255)',
      fontFamily: 'system-ui',
      border: '2px solid rgb(30,144,255)',
      borderRadius: '3px',
      fontSize: '15px',
      fontWeight: 'bold',
      cursor: 'pointer'
    };

    const paymentText = {
      position: 'absolute',
      top: '70px',
      left: '119px',
      fontSize: '19px',
      color: 'rgb(96,96,96)',
      textAlign: 'center',
      fontFamily: 'Trebuchet MS, Helvetica, sans-serif'
    };

    const closeBtn = {
      position: 'absolute',
      right: '20px',
      top: '22px',
      zIndex: '1',
      cursor: 'pointer',
      fontSize: '22px',
      fontWeight: 'bold'
    };

    return (
      <div style={overLayStyle}>
      <span style={closeBtn} onClick={hideOverlay}>&#215;</span>
      <p style={paymentText}>Select the amount to donate (EUR)</p>
      {payments.map((amount, index) => {
          return (
            <div style={inputStyle} key={index} className="radio">
            <label>
              <input
               name="payment"
               type="radio"
               value={amount}
               onClick={() => {this.selectedAmount(amount)}}
              /><span></span><span style={textStyle}>{amount}</span>
            </label>
            </div>
          )
      })}
      <button 
      style={payBtn}
      onClick={this.handlePay.bind(
      this,
      item.id,
      this.state.selectedAmount,
      item.currency
      )}
      >Pay
      </button>
      </div>
    );
  }
}
 
export default OverLay;
