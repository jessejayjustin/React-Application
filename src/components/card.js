import React, { useEffect, useRef, useState } from 'react';
import Input from './input';

const card = ({id, name, image, currency, currIndex, handlePayment, handleOverlay}) => {
  
  const [amt, setAmt] = useState(10);
  const [payments, setPayments] = useState([10,20,50,100,500]);

  const refs = useRef([React.createRef(), React.createRef()]);

  const overlayRef = useRef();
  const openOverlayRef = useRef();
  const closeOverlayRef = useRef();

  useEffect(() => {

    const handleOverlayOpen = () => {
      overlayRef.current.style.width = '100%';
      refs.current[0].current.checked = "true";
    }
    const handleOverlayClose = () => overlayRef.current.style.width = '0';

    openOverlayRef.current.addEventListener('click', handleOverlayOpen);
    closeOverlayRef.current.addEventListener('click', handleOverlayClose);

    return () => {
      openOverlayRef.current.removeEventListener('click', handleOverlayOpen);
      closeOverlayRef.current.removeEventListener('click', handleOverlayClose);
    }

  }, [])

  return (
    <div className="col-lg-5 col-md-9 col-sm-11 column">
      <div className="card">
        <div ref={overlayRef} className="overlay">
          <a ref={closeOverlayRef} className="float-right mr-3 mt-3 close-btn">x</a>
          <p className="title text-center">Select the amount to donate</p>
          <div className="btn-group d-flex justify-content-center">
            {
              payments.map((amount, index) => {
                return (
                  <Input 
                   className="inpt"
                   key={index}
                   Id={id}
                   index={index}
                   type="radio"
                   value={amount} 
                   currIndex={currIndex}
                   ref={refs.current[index]}
                   id={currIndex===id ? index : ""} 
                   name="customRadioInline1" 
                   onClick={() => setAmt(amount)}
                  />
                )
              })
            }
          </div>
          <button 
           className="btn btn-sm btn-outline-primary mt-5 d-flex mx-auto"
           disabled={currIndex===id ? false : true}
           onClick={handlePayment.bind(
            this,
            id,
            amt,
            currency
           )}
           >Pay
           </button>
        </div>
        <img className="img-fluid" src={require(`../../public/images/${image}`)} />
        <div className="card-body">
          <h5 className="card-title">{name}<a href="#" className="float-right"></a></h5>
          <button ref={openOverlayRef} className="btn btn-sm btn-outline-primary float-right" onClick={handleOverlay}>Donate</button>
        </div>
      </div>
    </div>
  )
}

export default card;





