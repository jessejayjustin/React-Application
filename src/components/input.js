import React from 'react';

function Input({Id, index, type, value, currIndex, id, name, onClick}, ref) {
  return (
    <div key={index} className="custom-control custom-radio custom-control-inline">
      <input 
        type={type} 
        ref={ref}
        id={id} 
        name={name}
        className="custom-control-input"
        value={value} 
        onClick={onClick}
        />
      <label className="custom-control-label" htmlFor={currIndex===Id ? index : ""}>{value}</label>
    </div>  
  )
}

const forwardRef = React.forwardRef(Input);

export default forwardRef;