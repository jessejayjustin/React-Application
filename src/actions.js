import { GET_CHARITIES, UPDATE_TOTAL_DONATIONS, UPDATE_MESSAGE, ERROR_MESSAGE } from "./action-types";
import { summaryDonations } from './helpers';

export function getCharities() {
    return dispatch => {
      const url = `http://localhost:3001/charities`;
      return fetch(url)
      .then(function(resp) { return resp.json(); })
      .then(function(data) {
        dispatch({ type: GET_CHARITIES, payload: data });
      }).catch(err => {
        dispatch({ type: ERROR_MESSAGE, payload: err });
      });
    };
};

export function getDonations() {
    return dispatch => {
      const url = `http://localhost:3001/payments`;
      return fetch(url)
      .then(function(resp) { return resp.json(); })
      .then(function(data) {
        dispatch({
          type: UPDATE_TOTAL_DONATIONS, 
          payload: summaryDonations(data.map((item) => (item.amount)))
        });
      }).catch(err => {
        dispatch({ type: ERROR_MESSAGE, payload: err });
      });
    };
};

export function handleAddPayment(id, amt, currency) {
    return dispatch => {
        return fetch(`http://localhost:3001/payments`, {
        method: 'POST',
        body: `{ "charitiesId": ${id}, "amount": ${amt}, "currency": "${currency}" }`,
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json"
        },
      }).then((response) => {
        response.json().then(data =>{
          dispatch({type: UPDATE_TOTAL_DONATIONS, payload: data.amount});
          dispatch({
            type: UPDATE_MESSAGE,
            payload: `Thanks for donating Â£ ${data.amount}`,
          });
          setTimeout(function() {
            dispatch({
              type: UPDATE_MESSAGE,
              message: ``,
            });
          }, 2000);
        })
      }).catch(err => {
          dispatch({ type: ERROR_MESSAGE, payload: err });
      });
    };
};