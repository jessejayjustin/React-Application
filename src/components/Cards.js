import React, { Component } from "react";
import styled from 'styled-components';
import OverLay from "./OverLay";

const Container = styled.div`
  background-color: white;
  width: 530px;
  height: 320px;
  position: relative;
  left: 115px;
  margin: 20px 20px;
  float: left;
  border-radius: 5px;
  -moz-box-shadow: 0px 3px 11px rgb(211,211,211);
  -webkit-box-shadow: 0px 3px 11px rgb(211,211,211);
  box-shadow: 0px 3px 15px rgb(211,211,211);
`;

export class Cards extends Component {
  constructor() {
    super();

    this.state = {
      selectedAmount: 10,
      active: null,
      activate: null
    };
    this.myVisibility = this.myVisibility.bind(this);   // Bind returns a new function, allowing you to pass in a this array and any number of arguments.
    this.active = this.active.bind(this);
    this.disable = this.disable.bind(this);
    this.btnDisabled = this.btnDisabled.bind(this);
    this.hideOverlay = this.hideOverlay.bind(this);
  }

  active(index) {
    if (this.state.active === index) {
      this.setState({active : null})
    } else {
      this.setState({active : index})
    }
  }

  myVisibility(index) {
    if (this.state.active === index) {
      return "visible";
    }
    return "hidden";
  }

  disable(item) {
    if (this.state.activate === item) {
      this.setState({activate : null})
    } else {
      this.setState({activate : item})
    }
  }

  btnDisabled(item) {
    if (this.state.activate === item) {
      return true;
    }
    return false;
  }

  hideOverlay() {
    //console.log("button clicked")
    this.setState({active : null})
    this.setState({activate : null})
  }

  render() {

    const imageStyle = {
      position: 'absolute',
      top: '0px',
      right: '0px',
      zIndex: '1',
      maxHeight:'100%',
      width: '530px',
      height: '245px',
      backgroundSize: 'cover',
      backgroundRepeat: 'no-repeat',
      backgroundPosition: 'center'
    };

    const btnStyle = {
      position: 'absolute',
      bottom: '23px',
      right: '25px',
      padding: '4px 9px',
      backgroundColor: 'rgb(255,255,255)',
      color: 'rgb(30,144,255)',
      fontFamily: 'system-ui',
      border: '2px solid rgb(30,144,255)',
      borderRadius: '3px',
      fontSize: '15px',
      fontWeight: 'bold',
      cursor: 'pointer'
    };

    const itemName = {
      position: 'absolute',
      bottom: '11px',
      left: '22px',
      fontSize: '17px',
      color: 'rgb(96,96,96)',
      fontFamily: 'Trebuchet MS, Helvetica, sans-serif'
    };

    const {charities, payments, handlePay} = this.props;
   
    return (
      <div className="app">
      {charities.map((item, index) => {
          return (
            <Container key={index} id="container">
            <img style={imageStyle} src={require(`../../public/images/${item.image}`)} alt="The image" className="img-responsive" />
            <p style={itemName}>{item.name}</p>
           
            <button style={btnStyle} onClick={() => { this.active(index); this.disable(index) }} disabled={this.btnDisabled(index)}>Donate</button> {/* onClick handler, invoking multiple functions */}
            
            <div style={{visibility:this.myVisibility(index)}}> {/* Using identifiers on map elements i can now activate/deactivate them all every time. By referencing each of them in a different way, then i can set the visibility property individually. */}
            <OverLay 
              payments={payments}
              item={item}
              handlePay={handlePay}
              hideOverlay={this.hideOverlay}
            />
            </div>
            </Container>
          )
      })}
      </div>
    );
  }
}

export default Cards;




