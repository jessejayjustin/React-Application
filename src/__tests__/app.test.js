import React from 'react';
import nock from 'nock';
import expect from 'expect';
import thunk from 'redux-thunk';
import fetchMock from 'fetch-mock'
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import { configure, mount, shallow, render } from 'enzyme';
import SmartComponent, { App } from '../App';
import { Card } from '../components/card';
import * as actions from '../actions';
import * as types from '../action-types';
import summaryDonations from '../helpers';
configure({ adapter: new Adapter() });

const props = {
  getCharities: jest.fn(),
  getDonations: jest.fn(),
  handleAddPayment: jest.fn()
}

const charities = [ 
  {
    "id": 1,
    "name": "Lorem ipsum dolor sit amet",
    "image": "hchi16.jpg",
    "currency": "EUR"
  },
  {
    "id": 2,
    "name": "Ut enim ad minim veniam",
    "image": "hchi2.jpg",
    "currency": "EUR"
  },
  {
    "id": 3,
    "name": "Neque porro quisquam est",
    "image": "hchi4.jpg",
    "currency": "EUR"
  }
];

describe('Component Rendering', () => {
  let wrapper, container, containerProp;

  beforeEach(() => {
    wrapper = shallow(<App charities={charities} {...props} />);
    container = wrapper.find('h3');
    containerProp = container.props();
  });

  it('renders <App> component', () => {
    expect(wrapper).toBeDefined();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a <div>', () => {
    expect(wrapper.find('div')).toHaveLength(3); 
  });

  it('should have a <div>', () => {
    expect(wrapper.find('p')).toHaveLength(1); 
  });

  it('should have a <div> with className container', () => {
    expect(wrapper.find('.container')).toHaveLength(1);
  });

  it('should have a <div> with className header', () => {
    expect(wrapper.find('.header')).toHaveLength(1);
  });

  it('should have a <div> with className row', () => {
    expect(wrapper.find('.row')).toHaveLength(1);
  });

  it('renders child component', () => {
    expect(wrapper.find('Card')).toBeDefined();
  });

  it('should have a <div> with properly className prop', () => {
    expect(containerProp.className).toEqual('text-center title');
  });

});

const middlewares = [thunk]; // add your middlewares like redux-thunk`
const mockStore = configureMockStore(middlewares);

describe('>>><SmartComponent/> --- REACT-REDUX (wrapping in <Provider>)',()=>{
  let store, wrapper;

  const initialState = {
    donations: 0,
    charities: [],
    message: '',
    error: null,
  }

  beforeEach(()=>{
    store = mockStore(initialState)
    wrapper = shallow(<Provider store={store}><SmartComponent/></Provider>)
  });

  it('+++ render the <SmartComponent/>', () => {
    expect(wrapper.find(SmartComponent).length).toEqual(1)
  });
});

describe('async actions', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('creates GET_CHARITIES_SUCCESS when fetching charities has been done', () => {
    nock('http://localhost:3001')
      .get('/charities')
      .reply(200, { charities: ['do something'] });

    const expectedActions = [
      { type: types.GET_CHARITIES_SUCCESS, payload: { charities: ['do something']  } }
    ]

    const store = mockStore({ charities: [] });
    store.dispatch(actions.getCharities()).then(() => {
      expect(store.getActions()).toEqual(expectedActions)
    });
  });
});

describe('async actions', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('creates UPDATE_TOTAL_DONATIONS when fetching donations has been done', () => {
    
    nock('http://localhost:3001')
      .get('/donations')
      .reply(200, {});

    const expectedActions = [
      { type: types.UPDATE_TOTAL_DONATIONS, payload: 0 }
    ];

    const store = mockStore({});
    store.dispatch(actions.getDonations()).then(() => {
      expect(store.getActions()).toEqual(expectedActions)
    });
  });
});
/*
describe('async actions', () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it('creates UPDATE_TOTAL_DONATIONS when post donations has been done', () => {
    fetchMock.getOnce('/donations', {
      body: { donations: ['do something'] },
      headers: { 'content-type': 'application/json' }
    })

    const expectedActions = [
      { type: types.FETCH_TODOS_REQUEST },
      { type: types.FETCH_TODOS_SUCCESS, body: { donations: ['do something'] } }
    ]
    const store = mockStore({ donations: [] })

    return store.dispatch(actions.handleAddPayment()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions)
    })
  })
})
*/
     





