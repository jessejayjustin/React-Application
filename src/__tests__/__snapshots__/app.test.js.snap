// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Rendering renders <App> component 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    <h3
      className="text-center title"
    >
      Donate to Save the Children Foundation
    </h3>
    <p
      className="msg"
    >
      All donations: Â£ 
    </p>
  </div>
  <div
    className="row pb-5 d-flex flex-row justify-content-center"
  >
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={0}
      image="hchi16.jpg"
      key="0"
      name="Lorem ipsum dolor sit amet"
    />
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={1}
      image="hchi2.jpg"
      key="1"
      name="Ut enim ad minim veniam"
    />
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={2}
      image="hchi4.jpg"
      key="2"
      name="Neque porro quisquam est"
    />
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={3}
      image="hchi3.jpg"
      key="3"
      name="Sed sed mollis justo"
    />
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={4}
      image="hchi15.jpg"
      key="4"
      name="Donec id molestie sapien"
    />
    <card
      className="child"
      currIndex={null}
      currency="EUR"
      handleOverlay={[Function]}
      handlePayment={[Function]}
      id={5}
      image="hchi12.jpg"
      key="5"
      name="Et sodales augue elit et dui"
    />
  </div>
</div>
`;
